Implement the history overlay mechanism

iamrecursion opened the pull request:
This mechanism is intended to be used for the forest implementation with pluggable storage backends. This forest will eventually replace the existing in-memory forest implementation, but for now is being developed in a separate portion of the tree to avoid breakage. We are aiming to merge portions of it incrementally to avoid mammoth PRs and increase visibility, and this mechanism being relatively self-contained makes it a good candidate for this.

This history mechanism is based on the following simplifying assumptions:

- A forest is generally used to store multiple independent trees and versions of each tree, but does not need to share between those independent trees.
- Historical versions are not important to persist, and can hence be handled entirely in memory.

The commit contains the basic functionality of this history storage and overlay mechanism, and tests that it functions correctly. It is almost entirely internal code, with little change to the public interface, and as a result temporarily uses `#[allow(dead_code)]` to keep the build warnings clean. These will be removed in future commits.

The interface for the `History` in this commit is a best effort at the design, and will likely need to undergo some minor changes to support concurrency as the shape of said concurrency in the forest itself is established.

This is the first bit of work toward #670. I would argue that this does not require a changelog entry as it is not user-facing.

## Checklist before requesting a review

- [x] Repo forked and branch created from `next` according to naming convention.
- [x] Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- [x] Relevant issues are linked in the PR description.
- [x] Tests added for new functionality.
- [x] Documentation/comments updated according to changes.

bobbinth commented on miden-crypto/src/merkle/smt/large_forest/history.rs:
```diff
@@ -0,0 +1,710 @@
+        root: Word,
+        nodes: NodeChanges,
+        leaves: LeafChanges,
+    ) -> Option<Delta> {
```
A couple of potential changes to this API:

- We could take a `MutationSet` instead of `root`, `nodes`, and `leaves` as for our specific use case, updates will be generated via mutation sets.
- I would also pass in a `version` so that we could associate a given mutation set with a version. This should make pruning of history a bit easier in the future.

Combining the above, I think the signature could look something like this:
```rust
pub fn add_version(
    &mut self,
    version: u32,
    updates: MutationSet<SMT_DEPTH, Word, Word>,
) -> Option<Delta> {
    ...
}
```

iamrecursion commented:
Okay, I have pushed a new commit that addresses all of the feedback except the one comment that I have left unresolved. What follows is a summary of [this](https://reilabs.slack.com/archives/C093G3C6MV2/p1764764288109159?thread_ts=1764663468.828509&cid=C093G3C6MV2) discussion on Slack.

It seems to me—and I may well be missing something—that we cannot construct the _reverse_ delta information passed to `add_version` with the _forward_ delta information in `MutationSet`. That type does not contain the previous value for any of the changes, which in the case of _overwrites_ (of both nodes and leaves) makes it insufficient as a source to construct the reverse delta information. 

Note that the `History` is a container purely for the overlays, and does not have any notion of the tree that it is overlaying.

If we want to make use of `MutationSet` there are three options in my mind:

1. Add the previous values to `MutationSet` at a memory cost, but this has good safety as the `MutationSet` is only constructed in one place and could store the prior values in the same traversal.
2. Require taking the `MutationSet` and an `impl SparseMerkleTree`. This is footgunny as you have to check at runtime whether the tree and the set actually correspond. I also have concerns whether implementing `SparseMerkleTree` on the individual trees in the forest's `Storage` will be able to be done performantly.
3. Require taking the `MutationSet` and some auxiliary data that boils down to the reversals  only where necessary, but this is very similar to the existing API. 

If all you meant was to provide a `MutationSet` API for the forest itself, this is a separate concern than the API to history, and is definitely planned.

bobbinth commented on miden-crypto/src/merkle/smt/history/mod.rs:
```diff
@@ -0,0 +1,923 @@
+        version_id: VersionId,
+        nodes: NodeChanges,
+        leaves: LeafChanges,
+    ) -> Result<Option<Delta>> {
```
A couple of questions:

- Why do we need to return the oldest version from here? i.e., why not make the return type just `Result<()>`?
- Why not replace `root`, `nodes`, and `leaves` with `MutationSet` as I mentioned in https://github.com/0xMiden/crypto/pull/679#discussion_r2579953876?

bobbinth reviewed (commented):
Looks good! Than you! Not a full review but I left some comments inline.

bobbinth commented on miden-crypto/src/merkle/smt/history/mod.rs:
```diff
@@ -0,0 +1,429 @@
+                return Err(HistoryError::VersionTooOld);
+            }
+        } else {
+            return Err(HistoryError::VersionTooOld);
```
I wonder if `VersionTooOld` is the right error here. For example, this could happen not because the version is too old, but because the tree just hasn't been updated recently. Not sure what's the best error here would be though.

bobbinth commented on miden-crypto/src/merkle/smt/history/mod.rs:
I wonder if the location of this file is correct. This is to be used specifically with `SmtForest` (or `LargeSmtForest` for now) - so, it seems like putting this under `src/merkle/smt/forest` (or `src/merkle/smt/large_forest` for now) may be a better option.

iamrecursion commented on miden-crypto/src/merkle/smt/large_forest/history/tests.rs:
Filed #699 to track.

drahnr commented on miden-crypto/src/merkle/smt/large_forest/history/mod.rs:
```diff
@@ -0,0 +1,430 @@
+    /// We use a [`VecDeque`] instead of a [`Vec`] or [`alloc::collections::LinkedList`] as we
+    /// estimate that the vast majority of removals will be the oldest entries as new ones are
+    /// pushed. This means that we can optimize for those removals along with indexing performance,
+    /// rather than optimizing for more rare removals from the middle of the sequence.
```
I don't quite understand how we would be able to do so, unless your interpretation is always one-delta-to-any-step, so the `Delta` is an accumulation of a set of reversions derived from i.e. a _forward_ `MutationSet`

iamrecursion replied:
Correct!

drahnr commented on miden-crypto/src/merkle/smt/large_forest/history/mod.rs:
```diff
@@ -0,0 +1,430 @@
+/// constructed by the history. Users should not be allowed to construct it directly.
+#[derive(Clone, Debug, PartialEq)]
+struct Delta {
+    /// The root of the tree in the `version` corresponding to this delta.
```
this means post applying the changes, this is the root we get? How do we ensure the delta is applied to the right "baseline"? Would it be good to have a second `Word`/digest to ensure it's targeting the correct tree state (read: runtime footgun check)?

iamrecursion replied:
Clarified the docs, but I want to avoid too many checks in the fast path for now. If this turns out to be a problem it can be added.

drahnr commented on miden-crypto/src/lib.rs:
```diff
@@ -49,6 +48,20 @@ pub type Map<K, V> = alloc::collections::BTreeMap<K, V>;
+/// By default, this is an alias for the [`alloc::collections::BTreeSet`]. However, when the
+/// `hashmaps` feature is enabled, this becomes an alias for hashbrown's HashSet.
+#[cfg(feature = "hashmaps")]
+pub type Set<V> = hashbrown::HashSet<V>;
```
note/nit: `hashbrown` is the `stdlib`'s actual implementation, we will move away from it soon~ish
